src/util/typed.rs

```diff
@@ -0,0 +1,2344 @@
+    log::trace!("sanitizing data: {:#?}", typed_data);
+    let data = typed_data
+        .as_object()
+        .ok_or_else(|| failure::err_msg("TypedData parameter must be a JSON Object"))?;
+
+    let primary_type = data
+        .get("primaryType")
+        .ok_or_else(|| failure::err_msg("primaryType must not present in the TypedData object"))?;
+    let primary_type: String = serde_json::from_value(primary_type.clone())
+        .map_err(|err| failure::format_err!("failed to deserialize primaryType: {}", err))?;
+
+    let domain = data
+        .get("domain")
+        .ok_or_else(|| failure::err_msg("domain must be present in the TypedData object"))?;
+    let domain = domain.clone();
+
+    let message = data
+        .get("message")
+        .ok_or_else(|| failure::err_msg("message must be present in the TypedData object"))?;
+    let message = message.clone();
+
+    let types = data
+        .get("types")
+        .ok_or_else(|| failure::err_msg("types must be present in the TypedData object"))?;
+    let types: HashMap<TypeName, Vec<SanitizedType>> = serde_json::from_value(types.clone())
+        .map_err(|err| failure::format_err!("failed to deserialize types: {}", err))?;
+
+    let data = SanitizedData {
+        types,
+        domain,
+        primary_type,
+        message,
+    };
+
+    log::trace!("sanitized data: {:#?}", data);
+
+    Ok(data)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::util::typed::{encode_type, find_type_dependencies, SanitizedData};
+    use bitcoin::util::misc::hex_bytes;
+    use std::collections::HashMap;
+
+    mod hash {
+        use crate::util::typed::hash;
+
+        fn do_test_hash(typed_data: &str, expected: &str) {
+            let input =
+                serde_json::from_str(typed_data).expect("failed to deserialize 'typed_data' json");
+            let got = hash(input).unwrap();
+            let got_base64 = base64::encode(&got);
+            assert_eq!(got_base64, expected);
+        }
+
+        #[test]
+        fn test_1() {
+            do_test_hash(
+                r#"{"types":{"EIP712Domain":[{"name":"name","type":"string"},{"name":"version","type":"string"},{"name":"chainId","type":"uint256"},{"name":"verifyingContract","type":"address"}],"Person":[{"name":"name","type":"string"},{"name":"wallet","type":"address"}],"Mail":[{"name":"from","type":"Person"},{"name":"to","type":"Person"},{"name":"contents","type":"string"}]},"primaryType":"Mail","domain":{"name":"Ether Mail","version":"1","chainId":1,"verifyingContract":"0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC"},"message":{"from":{"name":"Cow","wallet":"0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826"},"to":{"name":"Bob","wallet":"0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB"},"contents":"Hello, Bob!"}}"#,
+                r#"vmCa7jQ/s8Syjh355jL8pk/Prt4g8C6GJE793zCVe9I="#,
+            );
+        }
+        #[test]
+        fn test_2() {
+            do_test_hash(
```

Comment:
Are there any difference between test 2 and 3, 4 and 5 ?

Reply:
Forgot the dedup stage, yeah, will do.

Reply:
Deduplicated it, diff got a lot smaller :)